### PR TITLE
build: Switch to @loopback/build

### PR DESCRIPTION
**Pull request by raymondfeng**
See https://github.com/strongloop/loopback-next/pull/682

### CODE REVIEW

**Comment by raymondfeng** [comment]
@slnode test please

**Comment by virkt25** [comment]
I think `@loopback/build` needs to be released for CI to pass. Changes look good otherwise. Will review once CI is passing.

**Comment by raymondfeng** [comment]
I have released `@loopback/build` but ran into an issue to be fixed by https://github.com/strongloop/loopback-next/pull/689.

File: package.json [hunk 1]
```diff
@@ -8,20 +8,20 @@
   },
   "scripts": {
     "build": "npm run build:dist && npm run build:dist6",
-    "build:dist": "tsc --target es2017 --outDir dist",
-    "build:dist6": "tsc --target es2015 --outDir dist6",
-    "build:watch": "tsc --watch",
-    "clean": "rm -rf dist dist6",
+    "build:dist": "lb-tsc --target es2017 --outDir dist",
+    "build:dist6": "lb-tsc --target es2015 --outDir dist6",
+    "build:watch": "lb-tsc --watch",
```
Comment (bajtos) [inline package.json]: I am pretty unhappy with this change, all we did is change `tsc` to `lb-tsc`. The script is not simpler in any way that before. On the contrary, it adds one more level of abstraction that any developer using this package has to understand.

Before this change, anybody familiar with `tsc` could immediately understand what `npm run build:dist` is doing. With this change in place, they have to learn tricky details of how `lb-tsc` works.

For me, this is a net loss 👎

File: package.json [hunk 2]
```diff
@@ -8,20 +8,20 @@
...
+    "clean": "lb-clean dist dist6",
     "lint": "npm run prettier:check && npm run tslint",
     "lint:fix": "npm run prettier:fix && npm run tslint:fix",
-    "prettier:cli": "prettier \"**/*.ts\" \"**/*.js\"",
+    "prettier:cli": "lb-prettier \"**/*.ts\" \"**/*.js\"",
```
Comment (bajtos) [inline package.json]: Ditto.

For me, this is a net loss 👎

File: package.json [hunk 3]
```diff
@@ -8,20 +8,20 @@
...
     "prettier:check": "npm run prettier:cli -- -l",
     "prettier:fix": "npm run prettier:cli -- --write",
-    "tslint": "tslint -c tslint.full.json --project tsconfig.json --type-check",
+    "tslint": "lb-tslint -c tslint.build.json --project tsconfig.json",
```
Comment (bajtos) [inline package.json]: Ditto.

For me, this is a net loss 👎

File: package.json [hunk 4]
```diff
@@ -8,20 +8,20 @@
...
   "scripts": {
     "build": "npm run build:dist && npm run build:dist6",
-    "build:dist": "tsc --target es2017 --outDir dist",
-    "build:dist6": "tsc --target es2015 --outDir dist6",
-    "build:watch": "tsc --watch",
-    "clean": "rm -rf dist dist6",
+    "build:dist": "lb-tsc --target es2017 --outDir dist",
+    "build:dist6": "lb-tsc --target es2015 --outDir dist6",
+    "build:watch": "lb-tsc --watch",
+    "clean": "lb-clean dist dist6",
```
Comment (bajtos) [inline package.json]: I suppose this change is an improvement, although it makes me wonder why is it needed to specify `dist` and `dist6` arguments? I'd expect that `lb-clean` should take care of cleaning all artifacts produced by `lb-*` scripts. What is the benefit of `lb-clean` over calling `rimraf` (as a replacement of non-portable `rm -rf`)?

File: package.json [hunk 5]
```diff
@@ -8,20 +8,20 @@
...
+    "prettier:cli": "lb-prettier \"**/*.ts\" \"**/*.js\"",
     "prettier:check": "npm run prettier:cli -- -l",
     "prettier:fix": "npm run prettier:cli -- --write",
-    "tslint": "tslint -c tslint.full.json --project tsconfig.json --type-check",
+    "tslint": "lb-tslint -c tslint.build.json --project tsconfig.json",
     "tslint:fix": "npm run lint -- --fix",
     "prepublish": "npm run build",
     "pretest": "npm run clean && npm run build",
-    "test": "/bin/bash -c '[[ $(node -v) =~ [[:digit:]]+ ]] && [[ ${BASH_REMATCH[0]} -gt 6 ]]' && mocha dist/test || mocha dist6/test",
+    "test": "lb-dist mocha DIST/test",
```
Comment (bajtos) [inline package.json]: This is great and much more portable 👍

File: tsconfig.json [hunk 6]
```diff
@@ -1,18 +1,12 @@
...
-    "lib": ["es2017", "dom"],
-    "module": "commonjs",
-    "moduleResolution": "node",
-    "target": "es2017",
-    "outDir": "dist",
-    "sourceMap": true,
-    "declaration": true
-  },
-  "include": ["src", "test"]
+  "extends": "./node_modules/@loopback/build/config/tsconfig.common.json",
```
Comment (bajtos) [inline tsconfig.json]: Nice 👍

File: tsconfig.json [hunk 7]
```diff
@@ -1,18 +1,12 @@
...
+  "include": [
+    "src",
+    "test"
+  ],
+  "exclude": [
+    "node_modules/**",
+    "packages/*/node_modules/**",
+    "**/*.d.ts"
```
Comment (bajtos) [inline tsconfig.json]: Isn't this the default configuration of `tsc`? Why is it necessary to list these exclude items now, when the build was working well without them before?

File: tslint.json [hunk 8]
```diff
@@ -1,31 +1,4 @@
...
-    "no-var-keyword": true,
-    "triple-equals": [true, "allow-null-check", "allow-undefined-check"],
-    "typeof-compare": true,
-
-    // These rules make code maintenance easier
-    "no-default-export": true,
-    "prefer-const": true
-  }
+  "$schema": "http://json.schemastore.org/tslint",
+  "extends": ["./node_modules/@loopback/build/config/tslint.common.json"]
```
Comment (bajtos) [inline tslint.json]: Nice 👍 

I hope we will never run into the situation where `@loopback/build` ends up installed somewhere else in the node_modules hierarchy (e.g. in a parent directory).

File: package.json (hunk 4 above)
Comment (bajtos) [reply package.json]: (See https://www.npmjs.com/package/rimraf)

File: package.json (hunk 1 above)
Comment (raymondfeng) [reply package.json]: This has been cleaned up by https://github.com/strongloop/loopback4-extension-starter/pull/17. PTAL.

File: package.json (hunk 4 above)
Comment (raymondfeng) [reply package.json]: Same here.

**Comment by raymondfeng** [comment]
The main benefit to use `@loopback/build` is to reuse the dependencies and configurations so that modules opt in our build scripts don't have to maintain such things. For example, no need to add `devDependencies` such as typescript and tslint.

**Pull request merged**